From 196b3a11056c7967f17c539cf2ef1378b64106b5 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Wed, 10 Jul 2024 08:18:33 +0300
Subject: [PATCH] [FSW-2704] fs1: prefetch ppboot image

This patch adds functionality to prefetch ppboot image
from manifest before manifest validation starts. Basic
manifest checks and crc needs to be correct for image
to be prefetched. Only manifest validation flag can
be ignored for prefetch.

Change-Id: I581908b1cee483f132190b03a20d8d22559f82d3
---

diff --git a/app/boot_rom.c b/app/boot_rom.c
index 417e24f..c4f42c8 100644
--- a/app/boot_rom.c
+++ b/app/boot_rom.c
@@ -6,6 +6,7 @@
 
 #include <boot_rom.h>
 #include <security.h>
+#include <boot_flow.h>
 
 const int fs1_die_type = DIE_TYPE;
 
@@ -36,6 +37,8 @@
 
 	boot_rom_tpm_spi_setup();
 
+	boot_rom_prefetch_img(HL_FW_IMG_PPBOOT);
+
 	boot_rom_fake_fuse();
 
 	boot_rom_check_rma_enabled();
diff --git a/common/boot_desc.c b/common/boot_desc.c
index b992365..ec9f1eb 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -253,7 +253,7 @@
 {
 	boot_desc_update_ptrs(manifest_type);
 
-	return p_manifest_ctx->boot_desc_flag;
+	return !!(p_manifest_ctx->boot_desc_flag);
 }
 
 static bool boot_rom_is_valid_start_addr(struct boot_desc *desc)
diff --git a/common/boot_flow.c b/common/boot_flow.c
new file mode 100644
index 0000000..149eca9
--- /dev/null
+++ b/common/boot_flow.c
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <boot_rom.h>
+
+uint32_t boot_rom_prefetch_img(__unused enum hl_fw_img_type img_type)
+{
+	/* start the prefetch of ppboot or recovery images
+	 * In either case there can be multiple dmas involved
+	 * Maintain a state machine to do multiple dmas
+	 * It needs to decide which manifest it needs to do prefetch
+	 * from. If manifest crc or validation is invalid then it
+	 * needs to make sure to fetch image from other manifest.
+	 * Once the dma starts with one manifest, another function will
+	 * be implemented to complete it based on state machine.
+	 * return the manifest index from which it has started prefetch.
+	 */
+
+	return BOOT_ROM_SUCCESS;
+}
diff --git a/common/boot_flow.h b/common/boot_flow.h
new file mode 100644
index 0000000..0612c5c
--- /dev/null
+++ b/common/boot_flow.h
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+#ifndef _BOOT_FLOW_H_
+#define _BOOT_FLOW_H_
+
+uint32_t boot_rom_prefetch_img(enum hl_fw_img_type img_type);
+
+#endif /* _BOOT_FLOW_H_ */
